### ISSUE TITLE
[darc/maestro++] Notify on subscription update failures

### ISSUE THREAD

**Issue by mmitche** [dotnet/arcade#1387]
Maestro should notify (open issues) when there are subscription updates that fail to execute properly. This is just for internal infrastructure issues, and not failures in pull requests.

**Comment by jcagme** [comment]
@tmat informed that Roslyn was not getting arcade updates since there was a [PR](https://github.com/dotnet/roslyn/pull/32566/commits) which was created on 01/17 and has not been merged since there are merge conflicts. 

This PR has 19 commits so the part where Maestro updates a PR is working fine but there is no way to know if something is not working correctly. I quick checked the history of its subscription and found no related "errors".

I **think** (need to confirm) we stop trying to auto-merge when we find there are conflicts which is perfectly fine but also I think we should have a way to notify repo owners when a PR is in this state so action is taken and the PR can be manually merged.

**Comment by epananth** [comment]
Spoke to @riarenas and got the following update 

1) What kind of notification? - Create an issue in github/core-eng
2) How to differentiate between infrastructure error vs error in pull request? - Any error that happens before creating a PR is considered as infrastructure error and after the PR is created then its considered as PR error. 

In case of both the errors github issue is created. PR error can be assigned to the respective repo owners to take a look at.

**Comment by riarenas** [comment]
> How to differentiate between infrastructure error vs error in pull request? - Any error that happens before creating a PR is considered as infrastructure error and after the PR is created then its considered as PR error.

I'd add that in cases where there is a PR opened already, but we fail to update an existing PR with a new batch of dependencies, this also counts as one of the errors we want to report.

**Comment by epananth** [comment]
@riarenas  @mmitche 

I looked into the flow and this is what I propose. 

The code for creating and updating the PR exists in PullRequestActor. The idea is to catch the exception and create a github issue if the PR does not get created /updated. 

So I was planning to create a method to create a github pull request. 

**What will the method capture?**

	1) Error connecting to github while creating the PR / updating the PR
	2) Error connecting to Azdo 
	3) Any other error in calling the method CreatePullRequestAsync should go create a git hub issue. 
         4) Error while creating a new branch 
         5) Error while trying to commit to the branch


**What will the error contain?** 

The issue will be created in github/core-eng . The error will contain
  a) Source repo,
  b) Target repo and 
  c) Dependency update version( old and new) 
  d) Actual error details 

For Eg: Error while updating "x" dependency  from source repo to target repo. <Actual exception > HTTP 404 : Unable to connect to https://github.com/maestro-auth-test/maestro-test2

**Where should this method be called from?** 

   1)  In the PullRequestActor, there is a method called CreatePullRequestAsync. I was planning to call this method from the catch block. 
	2) Also the part where is says something bad happened but it did not create an exception this can be called from that place too. But here no exception is thrown so there will be no exception details in this issue. ( This might be a bad idea since we will not know what actually went wrong, but instead of just deleting the branch it will be nice to know the branch was deleted and no update went through) 

Please let me know your thoughts.

**Comment by epananth** [comment]
Spoke to @riarenas, we discussed that there are categories of errors in this which we cannot pin point at this particular time. So I am going to list the errors and categorize the errors and depending on the category it will create a different type of github issue( I have to think about this too) .

**Comment by mmitche** [comment]
Looks good to me. Where are you thinking to create issues for errors on PRs to AzDO repos? Maybe core-eng as a backup?

**Comment by epananth** [comment]
I categorized the list errors to start with, there might be more to this. 

Error | Error   category | Github | Notes
-- | -- | -- | --
Azdo   unable to connect | Azdo   Error | Single   Github issue | Need   to check what is the error we get when Azdo is down.        Eg   : Token expiration, Azdo 404 , 500       If   one of the subscriptions failed and an issue was created in github then for   the same error another issue should not be created.        Need   a way to look up the last few issues created in core-eng say last hour or so   then have to validate if we are creating an issue with the same info.       Not   sure if this is even possible.
Wrong parameters to the API request | Parameter Error | Single Github issue | There are many calls to different apis in PullRequestActor class.        Eg: Unable to create target branch, unable to find the target repo, unable   to find the valid subscription, wrong name of the repo, subscription, source   repo, target repo
Github   down | Github   Error | Single   Github issue | Eg:   Github API throttling        Should   create only one issue.
Any other errors | Unknown Error ( don’t know what type of error this should be) | Single Github issue | Before deleting the branch, it just creates an issue. (prob catch the   exception and add this in the description)


So I was thinking all the issues will be created in core-eng as default.

**Comment by mmitche** [comment]
> So I was thinking all the issues will be created in core-eng as default.
Works for me.

**Comment by epananth** [comment]
For Error diagnosis and capturing the error as github issue, we have 2 categories to look into

- 	SubscriptionUpdateHistory table - For non-batched updates - we can use the existing API -  /api/subscriptions/{id}/history
- 	RepositoryBranchUpdateHistory -  For Batched updates - we can use the existing API -  /api/repo-config/history

We can run a scheduled flow to check if there are any errors registered in the above tables. The scheduler can be configurable to run every hour or day. We already have some data in the table, but we need to improve these error msgs in them. 

**SubscriptionUpdateHistory** table has data which says it only some update went through. If we can add better error msgs then it will be useful for this scenario. Probably add more into like source, target repo, success message and error message.  Right now all the success and error message as logged under errorMessage column, we can update this to some column say Message that will hold a json with the success/ error message. 

**RepositoryBranchUpdateHistory** table definitely has better data which we can use. 

**Q's** - 
	

-  Is unable to merge because of merge conflict considered as an error? 
-  There are places in RepositoryBranchUpdateHistory where there is an error and after retry, the update is successful. I am not sure why we are not updating this record since the try was successful. I was thinking  we can add another column say RetryCount, we can keep count for the number of retries. We can do the same for the SubscriptionUpdateHistory. This will reduce duplication of the data. 
- Also not sure who is using this data. If we make changes to these tables what gets affected?

**Comment by riarenas** [comment]
Do you think the errors we are registering in the tables cover all the categories that you analyzed earlier? I like the idea of using these existing APIs as a source of truth for creating the issues, but depending of how much we record right now, we'll have to make sure we add enough details to make things actionable.

> Is unable to merge because of merge conflict considered as an error?
I think this is out of scope, we are only looking at informing on errors that happen before we can create or update the PR.

> There are places in RepositoryBranchUpdateHistory where there is an error and after retry, the update is successful. I am not sure why we are not updating this record since the try was successful. I was thinking we can add another column say RetryCount, we can keep count for the number of retries. We can do the same for the SubscriptionUpdateHistory. This will reduce duplication of the data.

I'm a bit lost here, isn't there a new record for the successful update? Are you suggesting we stop doing that, or are you asking why we do it like that?

> Also not sure who is using this data. If we make changes to these tables what gets affected?

Nothing that I'm aware of. We used to provide a darc command for just the subscription history, but removed it when we introduced the batched subscriptions and the repository history API.

**Comment by epananth** [comment]
I was looking into the data in SubscriptionUpdateHistory table and I could not find a lot of errors. So for the question do we know if this covers all the categories , I am honestly not sure. But I could see a variety of error in RepositoryUpdateHistory table. So after discussion with @riarenas,  it is nice to recreate the errors in SubscriptionUpdateHistory table same as RepositoryUpdateHistory table. In case it is some update to the table is missing, I am planning to add that as part of this issue. 

We are going to keep the columns in the table as is for now, reason being for some errors where we do not want to report it as a github issue it will be nice to see that errors in the table.

**Comment by epananth** [comment]
Design ideas | Pros | Cons
-- | -- | --
RepositoryBranchUpdateHistory   and    SubscriptionUpdateHistory       We need to make   some tweaks on the existing API.    Change all the   required params to optional params. Also add an extra parameter to track the   time from which time we need github issue has to be created.        This is support   existing feature plus new functionality. | Leverage the   existing API    Repurpose the   existing API which will work with and without params. | User might get   confused if the API per repo per branch vs get all the repo and branch   errors.    Do we really need   ppl to call this API? What if it fails when clicked multiple times?    Why do we have to   use API to call to the database when we are already in Maestro?
Create 2 new APIs   to get this info | Keep the APIs   separate.    No confusion if   its per repo/ per | This will have a   lot of redundancy.    Why do we have to   use API to call to the database when we are already in Maestro?
Run a scheduler   and get the info directly from the database.        **Probably we can   create a trigger API to trigger is manually. | Scheduler is   independent of API calls.    Cleaner and its   easier to call the db directly than calling the API. No tokens are needed for connecting to the database. | Will not be using   API

**Comment by epananth** [comment]
@riarenas I like the idea of directly connecting to the database to get the list of all the records and creating issue better. What do you think?

**Comment by riarenas** [comment]
I think it makes sense, I don't like the other two ideas that much either.

Do you think the errors we record right now are useful enough so that if we open an issue about a failure, it's something actionable? You mention that we'll keep adding new errors as we find them, which makes sense, but I also wonder if the existing errors are detailed enough that they are useful in the form of an issue, or if we need to change the data we are inserting into those tables.

A good start might be to find out which errors out of https://github.com/dotnet/arcade/issues/1387#issuecomment-573916862 are already captured, and if they have enough information for someone looking at a generated issue to know what to do about it.

**Comment by epananth** [comment]
Thanks @riarenas. I will go with the database approach. 

I added that to the design to keep this conversation open for future errors. Right now we have a good start for errors, I am going to use that. In future if there are any errors we can bucket it under one of the categories or may be even create a new category depending on the error.

**Comment by epananth** [comment]
After spending some considerable time with the Github search API, I figured that using error message / categories is not actually a great idea .I am currently using error message to search for a Github issue already created. Searching using a string gave me inconsistent results( sometimes I was able to find the string, sometimes I didn’t). Error message can be changed so, for the same msg we might end up creating multiple issues. 
 
@riarenas  and I met up and discussed about this. So we came up with a better approach. 
 
**Design :** 
 
Read the errors from the RepositoryBranchHistoryTable and create an issue in GitHub. Use a ReliableDictionary to store the Repo and Branch as key and the GitHub issue url as value. We are creating **only one issue** for Repository-branch combination. So if there are multiple failures for the same Repository-Branch combination, then the same issue will get updated if the issue is open. 
 
**What happens when another issue comes in for the same Repo branch combination?** 
 
Since we already have a reliable dictionary that contains the Repository-branch information, we check if there are any issue for this combination. If there already exists then update the GitHub issue else if there isn't any then create a new issue. 
 
**What will the issue contain?**
 
Subscription information if it exists, error message, method , repository and branch. 

cc : @mmitche @ChadNedzlek @alexperovich

**Comment by markwilkie** [comment]
I don't know a ton about reliable dictionaries.  So should I assume they persist "forever"?  I'm thinking about when the same error comes in "days" (or weeks?) later....

**Comment by epananth** [comment]
Reliable dictionaries are in-memory, as long as the cluster is up the data in the dictionary exists. If the same error comes in after days, it will update the GitHub issue if the issue is open else it will create a new GitHub issue. 

cc : @riarenas

**Comment by riarenas** [comment]
Not forever as in writing to a DB for this, but forever in that they will persist until we delete the cluster. It's what we already use to keep track of opened Dependency update PRs, which release pipelines we need to keep track of, etc...

They are in-memory, but also replicated across nodes and persisted to the node hard drives.

**Comment by alexperovich** [comment]
They aren't in-memory. They are stored on the hard-drive of each node and replicated "reliably" across the cluster. They are safe as long as the cluster is.

**Comment by alexperovich** [comment]
And, to be fair. Azure SQL databases are stored in service fabric's replicated data layer. So its super reliable.

**Comment by ChadNedzlek** [comment]
They will slowly grow forever, but hopefully the data is "small enough" that it won't be a problem, even if that happens.

Could we mark the issue with some stuff and search for that?  The alert opener thing for grafan shoves a "NotificationId:<guid>" and searches for that in the body to find things it opened, which seems to work fine (it's always found the issues it created and no others, to my knowledge, but it does limit the search by a specific label, so maybe that's helping).

That way we don't have 2 data sources we are trying to keep in sync (the issues themselves and then the issue) and will lose whenever we have to rebuild the cluster.  We've never stored non-transient data before.

**Comment by epananth** [comment]
For now the data isn't that big. We think its not going to be more than a few hundreds. . 

I used ErrorMessage (string)to find the issue, it was inconsistent. 2 out of 3 times it did find the issue. I tried putting the errorMessage in a commentMarker, but it did not find that either. I'm not sure if its the special characters or something else. I did remove all the special characters, but don't think it made any difference to it. So manipulating error message to find it the body, is not a great idea. But it is very fragile. 

I created an issue to back up reliable services for future work-> https://github.com/dotnet/arcade/issues/4933

As discussed, I will add log to get the number of issues in the dictionary each time the service runs. To keep an eye on the number.
